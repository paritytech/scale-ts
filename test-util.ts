/// <reference lib="deno.unstable"/>

import { assertEquals, assertThrows } from "https://deno.land/std@0.161.0/testing/asserts.ts"
import { assertSnapshot } from "https://deno.land/std@0.161.0/testing/snapshot.ts"
import { AnyShape, assert, DecodeBuffer, Shape, ShapeAssertError } from "./common/mod.ts"

const [lipsum, words, cargoLock] = ["lipsum.txt", "words.txt", "Cargo.lock"].map((fileName) => () =>
  Deno.readTextFile(fileName)
)
export const files = { lipsum: lipsum!, words: words!, cargoLock: cargoLock! }

export function testShape<T>(
  shape: Shape<T>,
  values: NoInfer<T>[] | Record<string, NoInfer<T> | (() => NoInfer<T>)>,
  async?: boolean,
): void
export function testShape<T>(
  shape: Shape<T>,
  values: T[] | Record<string, T | (() => T)>,
  async = false,
) {
  for (const key in values) {
    let value = values[key as never] as T | (() => T)
    const label = values instanceof Array
      ? Deno.inspect(value, {
        depth: Infinity,
        trailingComma: true,
        compact: true,
        iterableLimit: Infinity,
      })
      : key
    Deno.test(`${Deno.inspect(shape)} ${label}`, async (t) => {
      if (typeof value === "function") {
        value = (value as () => T)()
      }
      assert(shape, value)
      const encoded = async ? await shape.encodeAsync(value) : shape.encode(value)
      await assertSnapshot(t, encoded, { serializer: serializeU8A })
      const decodeBuffer = new DecodeBuffer(encoded)
      const decoded = shape.subDecode(decodeBuffer)
      assertEquals(decoded, value)
      assertEquals(decodeBuffer.index, encoded.length)
      assert(shape, decoded)
    })
  }
}

export function testInvalid(shape: AnyShape, values: unknown[] | Record<string, unknown | (() => unknown)>) {
  for (const key in values) {
    let value: unknown = values[key as never]
    const label = values instanceof Array
      ? Deno.inspect(value, {
        depth: Infinity,
        trailingComma: true,
        compact: true,
        iterableLimit: Infinity,
      })
      : key
    Deno.test(`${Deno.inspect(shape)} invalid ${label}`, async (t) => {
      if (typeof value === "function") {
        value = (value as () => unknown)()
      }
      await assertThrowsSnapshot(t, () => assert(shape as Shape<any>, value), ShapeAssertError)
    })
  }
}

function serializeU8A(array: Uint8Array) {
  return [...array].map((x) => x.toString(16).padStart(2, "0")).join("\n")
}

type NoInfer<T> = T extends infer U ? U : never
export function benchShape<T>(name: string, shape: Shape<T>, value: NoInfer<T>): void
export function benchShape<T>(name: string, shape: Shape<T>, value: T) {
  const encoded = shape.encode(value)
  Deno.bench(`- ${name} (${encoded.length}B) [encode] `, () => {
    shape.encode(value)
  })
  Deno.bench(`  ${name} (${encoded.length}B) [decode]`, () => {
    shape.decode(encoded)
  })
}

export async function assertThrowsSnapshot(
  t: Deno.TestContext,
  fn: () => unknown,
  ctor: abstract new(...args: any) => Error = Error,
) {
  try {
    fn()
  } catch (error) {
    if (!(error instanceof ctor)) {
      throw new Error("Expected " + ctor.name)
    }
    return await assertSnapshot(t, `${error.name}: ${error.message}`, { serializer: (x) => x })
  }
  throw new Error("Expected function to throw")
}

export { assertEquals, assertSnapshot, assertThrows }
