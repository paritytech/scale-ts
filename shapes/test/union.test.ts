import * as $ from "../../mod.ts"
import { metadata } from "../../mod.ts"
import { testInvalid, testShape } from "../../test-util.ts"

const $abc = $.withMetadata(
  metadata("$abc"),
  $.taggedUnion("_tag", [
    $.variant("A"),
    $.variant("B", $.field("B", $.str)),
    $.variant("C", $.field("C", $.tuple($.u32, $.u64))),
    $.variant("D", $.field("a", $.u32), $.field("b", $.u64)),
  ]),
)

const interestingU8s = {
  0: "Min",
  1: "Unit",
  2: "EvenPrime",
  28: "LargestPerfect",
  129: "FirstUninteresting",
  225: "LargestSquare",
  255: "Max",
} as const

const $interestingU8s = $.withMetadata(metadata("$.literalUnion(interestingU8s)"), $.literalUnion(interestingU8s))

const names = [
  "Ross",
  "Alisa",
  "Stefan",
  "Raoul",
  "James",
  "David",
  "Pierre",
] as const

const $names = $.withMetadata(metadata("$.literalUnion(names)"), $.literalUnion(names))

testShape($abc, [
  { _tag: "A" },
  { _tag: "B", B: "HELLO" },
  { _tag: "C", C: [255, 101010101n] },
  { _tag: "D", a: 101, b: 999n },
])

testShape($names, [...names])

testShape($interestingU8s, Object.values(interestingU8s))

testInvalid($abc, [
  null,
  { _tag: null },
  { _tag: "" },
  { _tag: "B" },
  { _tag: "B", B: null },
  { _tag: "D", a: 1 },
  { _tag: "D", b: 1n },
  { _tag: "D", a: -1, b: 1n },
  { _tag: "D", a: 1, b: -1n },
])
