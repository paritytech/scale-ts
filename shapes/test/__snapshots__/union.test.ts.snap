export const snapshot = {};

snapshot[`\$abc { _tag: "A" } 1`] = `00`;

snapshot[`\$abc { _tag: "B", B: "HELLO" } 1`] = `
01
14
48
45
4c
4c
4f
`;

snapshot[`\$abc { _tag: "C", C: [ 255, 101010101n ] } 1`] = `
02
ff
00
00
00
b5
4a
05
06
00
00
00
00
`;

snapshot[`\$abc { _tag: "D", a: 101, b: 999n } 1`] = `
03
65
00
00
00
e7
03
00
00
00
00
00
00
`;

snapshot[`\$.literalUnion(names) "Ross" 1`] = `00`;

snapshot[`\$.literalUnion(names) "Alisa" 1`] = `01`;

snapshot[`\$.literalUnion(names) "Stefan" 1`] = `02`;

snapshot[`\$.literalUnion(names) "Raoul" 1`] = `03`;

snapshot[`\$.literalUnion(names) "James" 1`] = `04`;

snapshot[`\$.literalUnion(names) "David" 1`] = `05`;

snapshot[`\$.literalUnion(names) "Pierre" 1`] = `06`;

snapshot[`\$.literalUnion(interestingU8s) "Min" 1`] = `00`;

snapshot[`\$.literalUnion(interestingU8s) "Unit" 1`] = `01`;

snapshot[`\$.literalUnion(interestingU8s) "EvenPrime" 1`] = `02`;

snapshot[`\$.literalUnion(interestingU8s) "LargestPerfect" 1`] = `1c`;

snapshot[`\$.literalUnion(interestingU8s) "FirstUninteresting" 1`] = `81`;

snapshot[`\$.literalUnion(interestingU8s) "LargestSquare" 1`] = `e1`;

snapshot[`\$.literalUnion(interestingU8s) "Max" 1`] = `ff`;

snapshot[`\$abc invalid null 1`] = `ShapeAssertError: value == null`;

snapshot[`\$abc invalid { _tag: null } 1`] = `ShapeAssertError: typeof value._tag !== "string"`;

snapshot[`\$abc invalid { _tag: "" } 1`] = `ShapeAssertError: value._tag: invalid tag`;

snapshot[`\$abc invalid { _tag: "B" } 1`] = `ShapeAssertError: !("B" in value)`;

snapshot[`\$abc invalid { _tag: "B", B: null } 1`] = `ShapeAssertError: typeof value.B !== "string"`;

snapshot[`\$abc invalid { _tag: "D", a: 1 } 1`] = `ShapeAssertError: !("b" in value)`;

snapshot[`\$abc invalid { _tag: "D", b: 1n } 1`] = `ShapeAssertError: !("a" in value)`;

snapshot[`\$abc invalid { _tag: "D", a: -1, b: 1n } 1`] = `ShapeAssertError: value.a < 0`;

snapshot[`\$abc invalid { _tag: "D", a: 1, b: -1n } 1`] = `ShapeAssertError: value.b < 0n`;
